galaxyproject/galaxy
# Adding new ARC job runner

This feature is a prototype created in the context of ESG. 

A new job runner - for submitting jobs to a remote ARC middleware endpoint. Behind the ARC endpoint is a regular batch system (slurm, htcondor etc). ARC endpoints are typically in front of designated grid infrastructures or HPC systems shared by other users. 

The use-case is a user that has access and compute-time to one or several HPC systems that are served by ARC. Instead of using the traditional command-line approach to send the jobs, Galaxy can be used. The authentication is handled by ARC, and the uploading of remote input datasets is also handled by ARC. The user would use the regular way of authenticating himself with ARC via tokens issued by one of the endpoints supported OIDC token providers. For this to be convenient the user should log into Galaxy using this supported OIDC token provider. Currently a WLCG IAM backend is in PR to social_core (https://github.com/python-social-auth/social-core/pull/820) which is one of the supported IdPs. The Galaxy server must be configured to use this oidc as a login-option by configuring the oidc_backends_config.yml. See details here: https://github.com/galaxyproject/galaxy/pull/16617

![Screenshot 2023-09-07 at 10 35 10](https://github.com/galaxyproject/galaxy/assets/22190352/d4fce9ee-c025-49c2-84ab-5fdafe6c3100)


Currently, the user must himself configure his user preferences with a single ARC endpoint to use. There is a feature request to ask for the possibility for the user to add as many endpoints are wanted: https://github.com/galaxyproject/galaxy/issues/16596
The form allowing users to add an ARC endpoint is currently merged into the usegalaxy-eu infrastructure-playbook: https://github.com/usegalaxy-eu/infrastructure-playbook/pull/879 . If you want to try this out, just have a look at this and apply the small additions to the relevant files. 

![Screenshot 2023-09-07 at 10 37 07](https://github.com/galaxyproject/galaxy/assets/22190352/04cfcdf4-fdd2-4086-8df5-3ac51f211bd3)


Currently also, the runner only works with a specific tool designated for ARC jobs. It has not yet been made ready for the using with toolbox tools. The arc-tool config file looks like this: 


```
<tool id="hello_arc" name="ARC hello">
  <description>is a simple hello world python script</description>
   <inputs>
   <param name="message" type="text" value="Hi from galaxy" label="Hello message"/>
   <param name="arcjob_exe" type="data" label="ARC executable"/>
   <param name="arcjob_cpuhrs" type="text" value="1" label="Job CPU hours"/>
   <param name="arcjob_memory" type="text" value="100" label="Job memory"/>-->
 </inputs>

  <outputs>
   <collection name="output" type="list"  label="ARC outputs">
     <discover_datasets pattern="__designation__"  directory="./" recurse="true" />
   </collection>
 </outputs>

  <help>
    TODO
  </help>
</tool>

```

![Screenshot 2023-09-07 at 10 37 43](https://github.com/galaxyproject/galaxy/assets/22190352/c3af26f7-3a65-436a-b40b-902df42b8fc6)


The tool expects that you upload an executable. In this case a simple bash script with the following contents: 
```
#!/bin/bash

/bin/echo "Hello from runscript to arcout1.txt" > ./arcout1.txt
/bin/echo "Hello from runscript to arcout2.txt" > ./arcout2.txt

/bin/echo "Running on compute node: $(hostname)" >> ./arcout1.txt
/bin/echo "Time is $(date)" >> ./arcout1.txt

/bin/echo "Running on compute node: $(hostname)" >> ./arcout2.txt
/bin/echo "Time is $(date)" >> ./arcout2.txt

/bin/echo "Sleeping for 60s " >> ./arcout1.txt
/bin/echo "Sleeping for 60s " >> ./arcout2.txt

>&1 /bin/echo "Some output to stdout from the runhello.sh executable."
/bin/sleep 60
```

But the executable could very well be some scientific code in C++ or whatever, as long as the ARC endpoint supports this software (this will be handled properly with so-called ARC runtime environment settings and matchmaking later). Currently this tool allows upload of a single executable and no other input-data just for testing purposes. In the future the job runner will support uploading of arbitrary number of input files from galaxy, in addition to remote input files (ARC collects these from the remote source). 

Once the job is done, all the output files created by the jobs executable and ARC itself are uploaded into Galaxy as expected. 

## How to test the changes?

  - For the galaxy admin: 
    - Install the ARC runner (in this PR) - the pyarcrest (ARC REST client) will be installed as a dependency. 
    - Configure the galaxy site to use the WLCG oidc backend (you might have to manually hack social_core backends as the PR mentioned above is not yet merged) - https://github.com/galaxyproject/galaxy/pull/16617 
    - Configure the galaxy site with the ARC test tool (see above)
    - Configure the galaxy site with the user_extra_preferences ARC distributed computing form: https://github.com/usegalaxy-eu/infrastructure-playbook/pull/879

  - For the user: 
      - Once the site is configured as above, you need to get authorized to get a token from WLCG IAM: https://wlcg.cloud.cnaf.infn.it/login Then log into galaxy via the WLCG login-button. 
      - Set your user-preferences to use an ARC endpoint that you have access to. You can use the test-ARC endpoint https://arctestcluster-slurm-el8-arc7-ce1.cern-test.uiocloud.no 
      - Create an executable file and upload it to the ARC tool. 
      - Submit the job - and sit back and relax. 


## License
